### PR TITLE
Easier JS Syntax 

### PR DESCRIPTION
**Pull request by Largo**
** Sorry, This is only  a draft. Any help to improve this is welcome**

Todo:
- Add Tests
- Refactor the code
- Extend == Operator to support comparisions with JS.eval for numbers

I'm using this in my projects: [Ruby Koans on the Browser](https://koans.idogawa.com)
[Ruby.wasm Quickstart](https://github.com/Largo/ruby.wasm-quickstart)

Using method missing for allowing easier JS syntax. #228 

- Type conversion from JS::Object to regular Ruby Objects
- == Comparison for nil and true 
- Make JSObjects enumerable to support each methods etc.
- Implement respond_to and allow seeing the available methods in IRB (only one level down so far Object1.method1.method2 does not work)

image

Shopping List Example code:

```ruby
JS.global.document.getElementById("spinner").style.display = "none"
        require "json"
        @document = JS.global.document
        @list = @document.getElementById("shopping-list")
        @input = @document.getElementById("new-item")
        @button = @document.getElementById("add-item")
        @storage = JS.global.localStorage

        def getItems()
          @storage.getItem("shoppingItems")
        end
        def updateStorage()
          items = []
            @document.getElementById("shopping-list").querySelectorAll("li").to_a.each do |item|
            label = item.querySelector("label").innerText
            isChecked = item.querySelector("[type=checkbox]").checked
            items.push({label:, isChecked:})
          end
          @storage.setItem("shoppingItems", JSON.generate(items))
        end

        def add_item(name, checked=false, shouldUpdateStorage=true)
          return if name.nil? || name&.strip&.empty?
          li = @document.createElement("li")
          checkbox = @document.createElement("input")
          checkbox.type = "checkbox"
          checkbox.checked = checked
          label = @document.createElement("label")
          label.innerText = name
          label.addEventListener("click") { |e| enable_editing(e.target) }
          li.appendChild(checkbox)
          li.appendChild(label)
          input = @document.createElement("input")
          input.type = "text"
          input.className = "edit-input"
          input.addEventListener("blur") { |e| update_item(e.target) }
          input.addEventListener("keypress") do |e|
            e.target.blur if e.key == "Enter"
          end
          li.appendChild(input)
          @list.appendChild(li)
          @input.value = ""
          updateStorage if shouldUpdateStorage
        end

        def enable_editing(label)
          input = label.nextSibling
          input.value = label.innerText
          label.style.display = "none"
          input.style.display = "block"
          input.focus
        end

        def update_item(input)
          label = input.previousSibling
          label.innerText = input.value.strip.empty? ? label.innerText : input.value
          label.style.display = "block"
          input.style.display = "none"
          updateStorage()
        end

        @button.addEventListener("click") { add_item(@input.value) }
        @input.addEventListener("keypress") do |e|
          add_item(@input.value) if e.key == "Enter"
        end
        @list.addEventListener("change") do |e|
        updateStorage
        end
        

        JSON.parse(getItems || "[]")&.to_a.each do |item|
          p item
          add_item(item["label"], item["isChecked"], false)
        end
```

### CODE REVIEW

File: packages/gems/js/lib/js.rb [hunk 1]
```diff
@@ -146,9 +158,88 @@ def new(*args)
...
+    # and change this from the method_missing
+    # make sure to run it in to _a
+
+    # JS::True
+    # Date to Ruby Date
+  end
+
+  # Support self == true instead of self == JS:True
+  alias_method :orig_eq, :==
+  def ==(other)
```
Comment (kateinoigakukun) [inline packages/gems/js/lib/js.rb:209]: How about comparing `self` with `other.to_js`?

File: packages/gems/js/lib/js.rb [hunk 2]
```diff
@@ -146,9 +158,88 @@ def new(*args)
...
+    JS.global[:Array].isArray(self) == JS::True
+  end
+
+  def self.static_to_rb(object)
+    return nil if self[sym] == JS::Null
+  end
+
+  def each(&block)
+    if block_given?
+      self.isJSArray ? self.to_a.each(&block) : Array.new(__props).each(&block)
```
Comment (kateinoigakukun) [inline packages/gems/js/lib/js.rb:231]: Hmm, I feel iterating on object properties is a little bit tricky. Do you have any use case for this behavior? I prefer simply delegating to `self.to_a.each`

**Review by kateinoigakukun — COMMENTED** [review]
Sorry, I don't have enough time to review all changes, but as a general feedback, would you mind splitting changes into smaller PRs and writing tests to cover new functionalities? It helps me a lot to review and I think it will give you some insights on edge cases.

**Comment by Largo** [comment]
I accidentaly closed this. I'm working on a new version that will be easier to review!
